File: src/MagnusAddon/note/vocabulary/vocabnote_forms.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from ankiutils import app
from ankiutils.app import col
from ex_autoslot import ProfilableAutoSlots
from note.note_constants import Mine, NoteFields
from note.notefields.comma_separated_strings_list_field_de_duplicated import MutableCommaSeparatedStringsListFieldDeDuplicated
from sysutils import ex_sequence, ex_str
from sysutils.collections.linq.q_iterable import QSet
from sysutils.lazy import Lazy
from sysutils.weak_ref import WeakRef, WeakRefable

if TYPE_CHECKING:
    from note.vocabulary.vocabnote import VocabNote

class VocabNoteForms(WeakRefable, ProfilableAutoSlots):
    def __init__(self, vocab: WeakRef[VocabNote]) -> None:
        self._vocab: WeakRef[VocabNote] = vocab
        field = MutableCommaSeparatedStringsListFieldDeDuplicated(vocab, NoteFields.Vocab.Forms)
        self._field: MutableCommaSeparatedStringsListFieldDeDuplicated = field
        weakrefthis = WeakRef(self)
        self._all_raw_set: Lazy[set[str]] = Lazy(lambda: set(weakrefthis()._field.get()))

        self._all_list: Lazy[list[str]] = field.lazy_reader(lambda: [ex_str.strip_brackets(form) for form in weakrefthis()._field.get()])
        self._all_set: Lazy[QSet[str]] = field.lazy_reader(lambda: QSet(weakrefthis()._all_list()))
        self._owned_forms: Lazy[set[str]] = field.lazy_reader(lambda: {weakrefthis()._vocab().get_question()} | {ex_str.strip_brackets(form) for form in weakrefthis()._all_raw_set() if form.startswith("[")})
        self._not_owned_by_other_vocab: Lazy[QSet[str]] = field.lazy_reader(lambda: weakrefthis().___not_owned_by_other_vocab())

    def is_owned_form(self, form: str) -> bool: return form in self._owned_forms()

    def all_list(self) -> list[str]: return self._all_list()
    def all_set(self) -> QSet[str]: return self._all_set()
    def all_raw_string(self) -> str: return self._field.raw_string_value()

    def all_list_notes(self) -> list[VocabNote]:
        return ex_sequence.flatten([app.col().vocab.with_question(form) for form in self.all_list()])

    def all_list_notes_by_sentence_count(self) -> list[VocabNote]:
        def prefer_more_sentences(vocab: VocabNote) -> int:
            return -vocab.sentences.counts().total

        return sorted(self.all_list_notes(), key=prefer_more_sentences)

    def not_owned_by_other_vocab(self) -> QSet[str]: return self._not_owned_by_other_vocab()

    def ___not_owned_by_other_vocab(self) -> QSet[str]:
        vocab_note = self._vocab()

        def is_not_owned_by_other_form_note(form: str) -> bool:
            return (app.col().vocab.with_question(form)
                    .where(lambda form_owning_vocab:
                           form_owning_vocab != vocab_note
                           and vocab_note.get_question() in form_owning_vocab.forms.all_set())
                    .none())
            # return not any(owner for owner in app.col().vocab.with_question(form)
            #                if owner != vocab_note and vocab_note.get_question() in owner.forms.all_set())

        return vocab_note.forms.all_set().where(is_not_owned_by_other_form_note).to_set()  # {form for form in vocab_note.forms.all_set() if not is_owned_by_other_form_note(form)}

    def without_noise_characters(self) -> list[str]:
        return [self._strip_noise_characters(form) for form in self.all_list()]

    @staticmethod
    def _strip_noise_characters(string: str) -> str:
        return string.replace(Mine.VocabPrefixSuffixMarker, "")

    def set_set(self, forms: set[str]) -> None: self.set_list(list(forms))

    def set_list(self, forms: list[str]) -> None: self._field.set(forms)

    def remove(self, remove: str) -> None:
        self._field.remove(remove)

        for remove_note in [voc for voc in col().vocab.with_question(remove) if self._vocab().get_question() in voc.forms.all_set()]:
            remove_note.forms.remove(self._vocab().get_question())

    def add(self, add: str) -> None:
        self._field.add(add)

        for add_note in [voc for voc in col().vocab.with_question(add) if self._vocab().get_question() not in voc.forms.all_set()]:
            add_note.forms.add(self._vocab().get_question())

    @override
    def __repr__(self) -> str: return self._field.__repr__()

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/VocabNoteForms.cs
using System.Collections.Generic;
using System.Linq;
using Compze.Utilities.SystemCE;
using JAStudio.Core.Note.CorpusData;

namespace JAStudio.Core.Note.Vocabulary;

public class VocabNoteForms
{
   readonly VocabNote _vocab;
   readonly NoteGuard _guard;
   List<string> _rawParts;
   readonly LazyCE<HashSet<string>> _allRawSet;
   readonly LazyCE<List<string>> _allList;
   readonly LazyCE<HashSet<string>> _allSet;
   readonly LazyCE<HashSet<string>> _ownedForms;
   readonly LazyCE<HashSet<string>> _notOwnedByOtherVocab;

   public VocabNoteForms(VocabNote vocab, VocabData? data, NoteGuard guard)
   {
      _vocab = vocab;
      _guard = guard;
      _rawParts = data?.Forms != null ? new List<string>(data.Forms) : [];
      _allRawSet = new LazyCE<HashSet<string>>(() => _rawParts.ToHashSet());
      _allList = new LazyCE<List<string>>(() => _rawParts.Select(StripBrackets).ToList());
      _allSet = new LazyCE<HashSet<string>>(() => _allList.Value.ToHashSet());
      _ownedForms = new LazyCE<HashSet<string>>(ComputeOwnedForms);
      _notOwnedByOtherVocab = new LazyCE<HashSet<string>>(ComputeNotOwnedByOtherVocab);
   }

   void InvalidateCaches()
   {
      _allRawSet.Reset();
      _allList.Reset();
      _allSet.Reset();
      _ownedForms.Reset();
      _notOwnedByOtherVocab.Reset();
   }

   HashSet<string> ComputeOwnedForms()
   {
      var owned = new HashSet<string> { _vocab.GetQuestion() };
      foreach(var form in _allRawSet.Value)
      {
         if(form.StartsWith("["))
         {
            owned.Add(StripBrackets(form));
         }
      }

      return owned;
   }

   HashSet<string> ComputeNotOwnedByOtherVocab()
   {
      bool IsNotOwnedByOtherFormNote(string form)
      {
         return !_vocab.Services.Collection.Vocab.Cache.WithQuestion(form)
                       .Any(formOwningVocab =>
                               formOwningVocab != _vocab &&
                               formOwningVocab.Forms.AllSet().Contains(_vocab.GetQuestion()));
      }

      return _vocab.Forms.AllSet()
                   .Where(IsNotOwnedByOtherFormNote)
                   .ToHashSet();
   }

   public bool IsOwnedForm(string form) => _ownedForms.Value.Contains(form);

   public HashSet<string> OwnedForms() => _ownedForms.Value;

   public List<string> AllList() => _allList.Value;
   public HashSet<string> AllSet() => _allSet.Value;
   public string AllRawString() => string.Join(", ", _rawParts);

   /// Returns all forms with brackets preserved (e.g. "[form]" stays as-is).
   public List<string> AllRawList() => _rawParts.ToList();

   public List<VocabNote> AllListNotes()
   {
      return _allList.Value
                     .SelectMany(form => _vocab.Services.Collection.Vocab.Cache.WithQuestion(form))
                     .ToList();
   }

   public List<VocabNote> AllListNotesBySentenceCount()
   {
      return AllListNotes()
            .OrderByDescending(vocab => vocab.Sentences.Counts().Total)
            .ToList();
   }

   public HashSet<string> NotOwnedByOtherVocab() => _notOwnedByOtherVocab.Value;

   public List<string> WithoutNoiseCharacters() => AllList().Select(StripNoiseCharacters).ToList();

   static string StripNoiseCharacters(string input) => input.Replace(Mine.VocabPrefixSuffixMarker, "");

   static string StripBrackets(string input) => input.Replace("[", "").Replace("]", "");

   public void SetSet(HashSet<string> forms)
   {
      SetList(forms.ToList());
   }

   public void SetList(List<string> forms) => _guard.Update(() =>
   {
      _rawParts = forms.Distinct().ToList();
      InvalidateCaches();
   });

   public void Remove(string remove)
   {
      _guard.Update(() =>
      {
         _rawParts = _rawParts.Where(item => item != remove).ToList();
         InvalidateCaches();
      });

      // Also remove from notes that have this vocab's question as a form
      var removeNotes = _vocab.Services.Collection.Vocab.Cache.WithQuestion(remove)
                              .Where(voc => voc.Forms.AllSet().Contains(_vocab.GetQuestion()))
                              .ToList();

      foreach(var removeNote in removeNotes)
      {
         removeNote.Forms.Remove(_vocab.GetQuestion());
      }
   }

   public void Add(string add)
   {
      _guard.Update(() =>
      {
         if(!_rawParts.Contains(add))
         {
            _rawParts.Add(add);
         }

         InvalidateCaches();
      });

      // Also add to notes that reference this form
      var addNotes = _vocab.Services.Collection.Vocab.Cache.WithQuestion(add)
                           .Where(voc => !voc.Forms.AllSet().Contains(_vocab.GetQuestion()))
                           .ToList();

      foreach(var addNote in addNotes)
      {
         addNote.Forms.Add(_vocab.GetQuestion());
      }
   }

   public override string ToString() => string.Join(", ", _rawParts);
}
